presto-cache/src/main/java/com/facebook/presto/cache/LocalRangeCacheManager.java

```diff
@@ -0,0 +1,461 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.cache;
+
+import com.facebook.airlift.log.Logger;
+import com.facebook.presto.spi.PrestoException;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeMap;
+import com.google.common.collect.TreeRangeMap;
+import io.airlift.slice.Slice;
+import org.apache.hadoop.fs.Path;
+
+import javax.annotation.PreDestroy;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.file.Files;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.Iterators.getOnlyElement;
+import static java.lang.StrictMath.toIntExact;
+import static java.nio.file.StandardOpenOption.APPEND;
+import static java.nio.file.StandardOpenOption.CREATE_NEW;
+import static java.util.Objects.requireNonNull;
+import static java.util.UUID.randomUUID;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+// 3 major TODOs for this class:
+// TODO: Make cache eviction based on cache size rather than file count; add evict count stats to CacheStats as well.
+// TODO: Make cache state persistent on disk so we do not need to wipe out cache every time we reboot a server.
+// TODO: File merge and inflight requests memory control.
+@SuppressWarnings("UnstableApiUsage")
+public class LocalRangeCacheManager
+        implements CacheManager
+{
+    private static final Logger log = Logger.get(LocalRangeCacheManager.class);
+
+    private static final String EXTENSION = ".cache";
+
+    private final ExecutorService cacheFlushExecutor;
+    private final ExecutorService cacheRemovalExecutor;
+
+    // a mapping from remote file `F` to a range map `M`; the corresponding local cache file for each range in `M` represents the cached chunk of `F`
+    private final Map<Path, CacheRange> persisted = new ConcurrentHashMap<>();
+    // a localing cache only to control the lifecycle of persisted
+    private final Cache<Path, Boolean> cache;
```

Comment:
For some reason I feel we are good with `Cache<Path, Void>` since this boolean is not used anyway.

Reply:
Unfortunately loading cache doesn't support nulls

```
com.google.common.cache.CacheLoader$InvalidCacheLoadException: CacheLoader returned null for key /var/folders/gm/gzznd_d56lb1vk2rhjhjm3zc0000gn/T/file1281856903816653633/data.

	at com.google.common.cache.LocalCache$Segment.getAndRecordStats(LocalCache.java:2314)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2154)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2044)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3952)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4871)
	at com.facebook.presto.cache.LocalRangeCacheManager.read(LocalRangeCacheManager.java:178)
	at com.facebook.presto.cache.LocalRangeCacheManager.get(LocalRangeCacheManager.java:138)
	at com.facebook.presto.cache.TestLocalRangeCacheManager.readFully(TestLocalRangeCacheManager.java:218)
	at com.facebook.presto.cache.TestLocalRangeCacheManager.testBasic(TestLocalRangeCacheManager.java:107)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:54)
	at org.testng.internal.InvokeMethodRunnable.run(InvokeMethodRunnable.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

Reply:
I wish there is a cache that only serves as a set =)